Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/CacheHandler.kt

```diff
@@ -0,0 +1,894 @@
+                CHUNK_CACHE_EXTENSION
+        )
+    }
+
+    private fun formatCacheFileName(originalFileName: String): String {
+        return String.format(
+                Locale.US,
+                CACHE_FILE_NAME_FORMAT,
+                originalFileName,
+                // AbstractFile expects all file names to have extensions
+                CACHE_EXTENSION
+        )
+    }
+
+    private fun formatCacheFileMetaName(originalFileName: String): String {
+        return String.format(
+                Locale.US,
+                CACHE_FILE_NAME_FORMAT,
+                originalFileName,
+                // AbstractFile expects all file names to have extensions
+                CACHE_META_EXTENSION
+        )
+    }
+
+    private fun createDirectories() {
+        if (!fileManager.exists(cacheDirFile)
+                && fileManager.create(cacheDirFile) == null) {
+            throw RuntimeException(
+                    "Unable to create file cache dir " + cacheDirFile.getFullPath())
+        }
+
+        if (!fileManager.exists(chunksCacheDirFile)
+                && fileManager.create(chunksCacheDirFile) == null) {
+            throw RuntimeException(
+                    "Unable to create file chunks cache dir " + chunksCacheDirFile.getFullPath())
+        }
+    }
+
+    private fun backgroundRecalculateSize() {
+        if (recalculationRunning.get()) {
+            // Already running. Do not use compareAndSet() here!
+            return
+        }
+
+        executor.submit { recalculateSize() }
+    }
+
+    private fun recalculateSize() {
+        var calculatedSize: Long = 0
+
+        if (!recalculationRunning.compareAndSet(false, true)) {
+            return
+        }
+
+        try {
+            val files = fileManager.listFiles(cacheDirFile)
+
+            for (file in files) {
+                if (fileManager.getName(file).endsWith(CACHE_META_EXTENSION)) {
+                    continue
+                }
+
+                calculatedSize += fileManager.getLength(file)
+            }
+
+            size.set(calculatedSize)
+        } finally {
+            recalculationRunning.set(false)
+        }
+    }
+
+    private fun trim() {
+        BackgroundUtils.ensureBackgroundThread()
+
+        val directoryFiles = fileManager.listFiles(cacheDirFile)
+        // Don't try to trim empty directories or just two files in it.
+        // Two (not one) because for every cache file we now have a cache file meta with some
+        // additional info.
+        if (directoryFiles.size <= 2) {
+            return
+        }
+
+        Logger.d(TAG, "trim() started")
+
+        // LastModified doesn't work on some platforms/phones
+        // (https://issuetracker.google.com/issues/36930892)
+        // so we have to use a workaround. When creating a cache file for a download we also create a
+        // meta file where we will put some info about this download: the main file creation time and
```

Comment:
This won't work, probably. I store a  boolean flag in the meta file that I set to true once the download is completed, chunks merged into the cache file and cache file is verified (md5 hashes match). I can't do that without the meta file.

Reply:
Right, what I mean is you can have an empty file be your meta file; if the meta file exists, that boolean isDownloaded flag is effectively false. If the meta file doesn't exist, the flag is true.

Reply:
Well, yes, this will probably work but still I don't really want to change this code because I've been testing it for weeks and if I change it I will have to test it again. There are even tests and shit. Are there any other reasons to change it other than it been kinda complex (it's not really  that complex, it's just two variables in a file that I read/update and if I can't access the file for some reason I just delete both so the file can be redownloaded)? Because this reason is not really motivating enough to change it.